dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/search/SearchIncludeFragment.java

```diff
@@ -755,22 +750,41 @@ public boolean hasValidFilterQueries() {
         return friendlyNames;
     }
 
-    public String getNewSelectedTypes(String typeClicked) {
-        List<String> newTypesSelected = new ArrayList<>();
-        for (String selectedType : selectedTypesList) {
-            if (selectedType.equals(typeClicked)) {
-
-            } else {
-                newTypesSelected.add(selectedType);
-            }
-
-        }
-        if (selectedTypesList.contains(typeClicked)) {
-
+    public boolean selectedTypesContainsDataverse() {
+        return selectedTypesList.contains(SearchObjectType.DATAVERSES);
+    }
+    public boolean selectedTypesContainsDataverseOnly() {
+        return selectedTypesList.size() == 1 && selectedTypesList.get(0) == SearchObjectType.DATAVERSES;
+    }
+    
+    public boolean selectedTypesContainsDataset() {
+        return selectedTypesList.contains(SearchObjectType.DATASETS);
+    }
+    public boolean selectedTypesContainsDatasetOnly() {
+        return selectedTypesList.size() == 1 && selectedTypesList.get(0) == SearchObjectType.DATASETS;
+    }
+    
+    public boolean selectedTypesContainsFile() {
+        return selectedTypesList.contains(SearchObjectType.FILES);
+    }
+    public boolean selectedTypesContainsFileOnly() {
+        return selectedTypesList.size() == 1 && selectedTypesList.get(0) == SearchObjectType.FILES;
+    }
+    
+    public String getNewSelectedTypes(String typeClickedString) {
+        SearchObjectType typeClicked = SearchObjectType.fromSolrValue(typeClickedString);
+        
+        List<SearchObjectType> newTypesSelected = new ArrayList<>(selectedTypesList);
+        
+        if (newTypesSelected.contains(typeClicked)) {
+            newTypesSelected.remove(typeClicked);
         } else {
             newTypesSelected.add(typeClicked);
         }
-        return StringUtils.join(newTypesSelected, ":");
+        
+        return newTypesSelected.stream()
+                .map(t -> t.getSolrValue())
+                .collect(Collectors.joining(":"));
```

Comment:
As for extraction to specialized class - I could use `SearchForTypes` object instead of `selectedTypesList` and extend it with methods `contains(SearchObjectType)`, `containsOnly(SearchObjectType)`, `toggleType(SearchObjectType)` then on xhtml use it like `SearchIncludeFragment.selectedTypes.contains(searchObjectTypeEnum.DATAVERSES)`.
What do you think?

Reply:
Yes (alternatively – if I'm not wrong – we could use some kind of interface with default methods for this ```…Fragment``` classes).